Make placements opaque

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #169867
* #172566
* __->__ #171482



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela

angelayi committed:
Make placements opaque

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -1805,7 +1806,7 @@ def check_user_input_output(flat_values: list[Any], error_type: UserErrorType) -
 
     def is_supported_type(val: Any) -> bool:
-        return isinstance(val, tuple(supported_types))
+        return isinstance(val, tuple(supported_types)) or is_opaque_type(type(val))
```
this path is for export flattener?

angelayi replied:
yes!

zou3519 commented on torch/_dynamo/variables/script_object.py:
```diff
@@ -218,6 +219,28 @@ def call_method(
+
+            if name == "__setattr__":
+                method(*args, **kwargs)
+                return self.value  # pyrefly: ignore[bad-return]
```
what about custom setattr? I'd expect us to ban that too

zou3519 commented on torch/_dynamo/variables/script_object.py:
```diff
@@ -218,6 +219,28 @@ def call_method(
+                method(*args, **kwargs)
+                return self.value  # pyrefly: ignore[bad-return]
+
+            constant_val = method(*args, **kwargs)
```
At a high level, we need to implement settattr because we are dynamo tracing through the constructor of the Opaque object?

angelayi replied:
Ah no sorry 😅 I saw the PlacementVariable had a setattr implementation and followed it. Seems like it is needed due to a hard-to-repro internal bug (https://github.com/pytorch/pytorch/pull/118668)?

zou3519 replied:
I noticed that we are calling `method(*args, **kwargs)` where (args, kwargs) are the real python values. Generally we don't allow that in Dynamo, because we don't know if the method is safe to execute (that is, if it has any side effects, like printing or appending to a global list).

There's a couple of ways to go about this:
1) We say that we're going to assume methods on Opaque values can be run with real values at compilation time.
2) We have Dynamo trace through the method via VariableTrackers.
3) We have some user-based annotation API that says which methods are safe to be run with real values and which aren't safe (and need graph breaks)

(3) is sounding a lot like the API in the PR above the stack

angelayi replied:
> Generally we don't allow that in Dynamo, because we don't know if the method is safe to execute (that is, if it has any side effects, like printing or appending to a global list).

I think that makes sense to me!

Originally my thinking was just focused on the graph itself -- since value types are specialized on and directly put in the graph, any in-place modification to it would result in the next usage of the value obj to have the modified inplace obj put into the graph. Like if you "in place modify" an integer, the next time you use it it would just be the new integer. 

But I didn't consider that these methods could have global size effects. I think it's fair to have a similar model to the reference types where you have to specify which methods/attributes to allow.

zou3519 replied:
If we're going to have a similar model, the question is if it is exactly the same model or a different one

angelayi commented on test/distributed/tensor/test_dtensor_compile.py:
```diff
@@ -957,7 +909,7 @@ def fn(x):
         mesh = DeviceMesh(self.device_type, torch.arange(self.world_size))
 
-        placements = PytreeTuple(Shard(0))
+        placements = [Shard(0)]
```
Ah I don't think there is a valid test that shows this yet, since from_local is represented as a closure. I did add an expecttest in https://github.com/pytorch/pytorch/pull/169867 which shows the placements in the graph though!

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"

This PR registers placements as opaque type through `_register_placements_as_opaque` but not globally (so that we don't break things). 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"

This PR registers placements as opaque type through `_register_placements_as_opaque` but not globally (so that we don't break things). 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"

This PR registers placements as opaque type through `_register_placements_as_opaque` but not globally (so that we don't break things). 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"

This PR registers placements as opaque type through `_register_placements_as_opaque` but not globally (so that we don't break things). 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

zou3519 commented on test/distributed/tensor/test_dtensor_compile.py:
```diff
@@ -903,7 +855,7 @@ def fn(x):
 
     def test_dynamo_to_local_grad_placements_sequence(self):
-        placements = PytreeTuple([Shard(0)])
+        placements = [Shard(0)]
```
> This PR registers placements as opaque type through _register_placements_as_opaque but not globally (so that we don't break things).

What can we break?

angelayi replied:
sorry that comment is outdated.. we actually register placements as opaque globally here

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi commented:
@pytorchbot merge

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi commented:
@pytorchbot merge

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi commented:
@pytorchmergebot revert -m "internal test failure"

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

izaitsevfb commented:
@pytorchbot revert -m "breaking internal conveyor, see D91862782" -c ghfirst

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

angelayi committed:
Update on "Make placements opaque"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]
